File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/candidate/AbstractNumCandidateCalculator.java
package apeman_core.features_extraction.calculators.candidate;

import apeman_core.features_extraction.calculators.BaseMetricsCalculator;
import apeman_core.utils.CandidateUtils;
import apeman_core.utils.MethodUtils;
import com.intellij.psi.JavaRecursiveElementVisitor;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiStatement;
import org.jetbrains.research.groups.ml_methods.utils.ExtractionCandidate;

import java.util.ArrayList;

abstract class AbstractNumCandidateCalculator extends BaseMetricsCalculator {

    private ArrayList<ExtractionCandidate> candidates;

    public AbstractNumCandidateCalculator(ArrayList<ExtractionCandidate> candidates) {
        this.candidates = new ArrayList<>(candidates);
    }

    public class CandidateVisitor extends JavaRecursiveElementVisitor {
        protected int methodNestingDepth = 0;
        ArrayList<ExtractionCandidate> methodCandidates;
        ArrayList<Integer> counts = new ArrayList<>();
        boolean isInsideMethod = false;

        @Override
        public void visitMethod(PsiMethod method) {
            if (methodNestingDepth == 0) {
                methodCandidates = CandidateUtils.getCandidatesOfMethod(method, candidates);
                initCounters();
                isInsideMethod = true;
            }

            methodNestingDepth++;
            super.visitMethod(method);
            methodNestingDepth--;

            if (methodNestingDepth == 0 && !MethodUtils.isAbstract(method)) {
                for (int i = 0; i < methodCandidates.size(); i++) {
//                    postMetric(methodCandidates.get(i), getCounterForCand(i));
                }
                candidates.removeAll(methodCandidates);
                isInsideMethod = false;
            }
        }

        protected void initCounters() {
            counts.clear();
            methodCandidates.forEach(elem -> counts.add(0));
        }

        protected int getCounterForCand(int i) {
            return counts.get(i);
        }

        protected void updateCounters() {
            for (int i = 0; i < methodCandidates.size(); i++) {
                updateCounter(i);
            }
        }

        protected void updateCounter(int i) {
            if (methodCandidates.get(i).isInCandidate()) {
                counts.set(i, counts.get(i) + 1);
            }
        }

        @Override
        public void visitStatement(PsiStatement statement) {
            CandidateUtils.checkStartOfCandidates(statement, methodCandidates);
            super.visitStatement(statement);
            CandidateUtils.checkEndOfCandidates(statement, methodCandidates);
        }
    }
}

File: ApemanPlugin/src/main/kotlin/apeman_core/features_extraction/calculators/candidate/AbstractNumCandidateCalculator.kt
package apeman_core.features_extraction.calculators.candidate

import apeman_core.base_entities.ExtractionCandidate
import apeman_core.base_entities.FeatureType
import apeman_core.features_extraction.calculators.BaseMetricsCalculator
import apeman_core.utils.CandidateUtils
import apeman_core.utils.MethodUtils
import com.intellij.psi.JavaRecursiveElementVisitor
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiStatement

import java.util.ArrayList

abstract class AbstractNumCandidateCalculator(candidates: List<ExtractionCandidate>, feature: FeatureType) : BaseMetricsCalculator(candidates, feature) {

    open inner class CandidateVisitor(
            val methodCandidates: List<ExtractionCandidate>
    ) : JavaRecursiveElementVisitor() {
        var methodNestingDepth = 0
        var counts: MutableList<Int> = ArrayList()
        var isInsideMethod = false

        override fun visitMethod(method: PsiMethod) {
            if (methodNestingDepth == 0) {
                initCounters()
                isInsideMethod = true
            }

            methodNestingDepth++
            super.visitMethod(method)
            methodNestingDepth--

            if (methodNestingDepth == 0 && !MethodUtils.isAbstract(method)) {
                for ((i, cand) in methodCandidates.withIndex()) {
                    results.set(cand, firstFeature, getCounterForCand(i).toDouble())
                }
                isInsideMethod = false
            }
        }

        protected open fun initCounters() {
            counts.clear()
            repeat(methodCandidates.size) { counts.add(0) }
        }

        protected open fun getCounterForCand(i: Int) = counts[i]

        protected open fun updateCounters() = methodCandidates.indices.forEach { updateCounter(it) }
        protected open fun updateCounter(i: Int) {
            if (methodCandidates[i].isInCandidate) {
                counts[i]++
            }
        }

        override fun visitStatement(statement: PsiStatement) {
            CandidateUtils.checkStartOfCandidates(statement, methodCandidates)
            super.visitStatement(statement)
            CandidateUtils.checkEndOfCandidates(statement, methodCandidates)
        }
    }
}
